algorithms/kernel/dtrees/forest/classification/oneapi/cl_kernels/df_batch_classification_kernels.cl

```diff
@@ -0,0 +1,573 @@
+        // nodeList will be updated with split attributes in this kernel
+        // spliInfo will contain node impurity and mean
+        const int nProp              = HIST_PROPS;                  // classes histogram
+        const int nNodeProp          = NODE_PROPS;                  // num of node properties in nodeList
+        const int nImpProp           = IMPURITY_PROPS + HIST_PROPS; // impurity + node classes histogram
+        const int local_id           = get_local_id(0);
+        const int sub_group_local_id = get_sub_group_local_id();
+        const int sub_group_size     = get_sub_group_size();
+        const int nodeIdx            = get_global_id(1);
+        const int nodeId             = nodeIndices[nodeIndicesOffset + nodeIdx];
+        const int leafMark           = -1;
+
+        const int local_size         = get_local_size(0);
+        const int n_sub_groups       = local_size / sub_group_size; // num of subgroups for current node processing
+        const int sub_group_id       = local_id / sub_group_size;
+        const int max_sub_groups_num = 16;                                                     //replace with define
+        const int bufIdx             = get_global_id(1) % (max_sub_groups_num / n_sub_groups); // local buffer is shared between 16 sub groups
+
+        const int rowsOffset = nodeList[nodeId * nNodeProp + 0];
+        const int nRows      = nodeList[nodeId * nNodeProp + 1];
+
+        // each sub group will process 16 bins and produce 1 best split for it
+        // expect maximum 16 subgroups only for each node
+        const int maxBinsBlocks = 16;
+        __local algorithmFPType bufI[maxBinsBlocks]; // storage for impurity decrease
+        __local int bufS[maxBinsBlocks * nNodeProp]; // storage for split info
+
+        int valNotFound     = 1 << 30;
+        int curFeatureValue = leafMark;
+        int curFeatureId    = leafMark;
+
+        const algorithmFPType minImpDec = (algorithmFPType)-1e30;
+        algorithmFPType curImpDec       = minImpDec;
+
+        nodeList[nodeId * nNodeProp + 2] = curFeatureId;
+        nodeList[nodeId * nNodeProp + 3] = curFeatureValue;
+        nodeList[nodeId * nNodeProp + 4] = nRows;
+
+        algorithmFPType mrgN = (algorithmFPType)nRows;
+
+        algorithmFPType mrgLN = (algorithmFPType)0;
+
+        algorithmFPType bestLN        = (algorithmFPType)0;
+        algorithmFPType mrgCls[nProp] = { (algorithmFPType)0 };
+
+        algorithmFPType imp = (algorithmFPType)1;
+
+        int totalBins = 0;
+
+        // totalBins is calculated by each subgroup
+        for (int featIdx = sub_group_local_id; featIdx < nSelectedFeatures; featIdx += sub_group_size)
+        {
+            int featId = selectedFeatures[nodeId * nSelectedFeatures + featIdx];
+            int nBins  = binOffsets[featId + 1] - binOffsets[featId];
+            totalBins += sub_group_reduce_add(nBins);
+        }
+
+        totalBins = sub_group_broadcast(totalBins, 0);
+
```

Comment:
Vlad mentioned that "reduce" is faster than "broadcast"

Reply:
Yes, you can make one single reduce after the loop, since you don't need intermediate values of aggregated totalBins inside.

Reply:
Will update it during hotspots analysis